elastic/kibana
# [SIEM][Security Solution][Endpoint] Endpoint Artifact Manifest Management + Artifact Download and Distribution

## Summary
This PR manages the entire lifecycle of exception list artifacts which will be delivered to the Elastic Endpoint. Major features include:

- A periodic task runner that
  - reads the current state of endpoint exception lists
  - builds artifacts for each supported OS, ~compresses them~ (compression TBD in next PR), and computes hashes
  - compares the hashes to the ones contained in the last-dispatched manifest
  - writes the artifacts to a Saved Object in ES  if new
  - builds a new manifest, removing outdated artifacts, and adding new ones
  - dispatches that manifest to the endpoints by writing to the appropriate Datasource config
  - and finally, commits the manifest to a Saved Object in ES if the dispatch was successful

- A new Kibana endpoint for downloading the artifacts
  - with authorization via Fleet API token
  - in-memory caching to provide for quicker responses
  - quick ES lookup based on a stable, reproducible doc id if cache lookup fails
  - the URL to each artifact is included in the manifest that is dispatched to the endpoints

### Properties

**Eventual Consistency from `lists` to `ingest_manager`**
The feature has been designed to run seamlessly in a multi-Kibana environment. The manifest SO that is committed is intended as a kind of "ack" to indicate that a manifest has been dispatched (written to ingest manager datasource). In the event of an unexpected server crash, it's possible that a manifest can be dispatched more than once, however it should never be possible for a manifest change to be lost.

There is one potential edge case... on the callback for datasource create, we're unable to verify that the change is actually committed at the ingest manager layer. A crash after we return could result in a manifest update being lost. **Revisit this?**

**Consistency of the Artifact Manifest**
The TaskManager plugin is leveraged for maintenance of the artifacts and the manifest commit records. This should ensure that only one task is running at once, however we do also have the ingestManager callback that runs when a datasource is created. In order to prevent race conditions and therefore consistency issues on the manifest, we utilize the `version` that is returned using the SavedObjectsClient APIs. If two clients attempt to make simultaneous updates against the same base version, a conflict (409) will be encountered, and the manifest will be updated on the next task run (within 60 seconds) if necessary.

**Scale Considerations**
The artifact manifest is now sent to the endpoint via the config/policy mechanism. The manifest can only be updated every 60 seconds, with the exception of when a new datasource is created, which could result in an out-of-band manifest update.

Artifacts are saved using a pre-calculated document ID, which encapsulates the schema version, operating system, artifact identifier, and fingerprint (a sha256 hash). Though downloads are currently managed through a Kibana API, we try to make the lookup as quickly as possible by utilizing a direct `get` by ID, and by utilizing an in-memory FIFO cache per client.


### Testing
1. Create an Endpoint config
2. Create the endpoint exception list:
```
cd x-pack/plugins/lists/server/scripts
./post_exception_list.sh exception_lists/new/exception_list_agnostic.json
```
3. Create an endpoint exception list item:
```
cd x-pack/plugins/lists/server/scripts
./post_exception_list_item.sh ./exception_lists/new/exception_list_item_agnostic.json
```
4. Navigate to the policy page and verify that `artifact_manifest` is also now contained in the Endpoint datasource
![image](https://user-images.githubusercontent.com/611653/85974058-7959f300-b9a2-11ea-9de9-048d90f4258e.png)


### To Do
- [x] create task to run artifact packager periodically
- ~add manifest endpoint~
- [x] add manifest generation
- [x] inject manifest into endpoint config (dispatch)
- [x] add download endpoint
- [x] in-memory cache of available downloads
- ~etag to cache manifest download~
- ~digitally sign manifest using RSA keypair~
- [x] authentication/authorization to APIs using api key
- [x] add io-ts schema validation
- [x] add api integration tests

### To be addressed in follow-up PR
- [x] compress the artifacts
- [x] apply new manifest schema
- [x] datasource update race prevention (BLOCKED: https://github.com/elastic/kibana/issues/69992)
- [x] prefer `async/await` to `then` (SO client abstractions should return `null` to avoid try/catch when possible)
- [x] fix `@ts-ignore` type errors
- [x] runtime type checking improvements (prefer `if(someType.is(...))` to `as`)
- [x] look for opportunities to use array.map, array.reduce, etc instead of for loops
- [ ] limit cache size
- [x] address PR comments
- [x] Fix paging bug: https://github.com/elastic/kibana/pull/67707/commits/f843f7066429beac611de125bdcac59076e078b3#diff-21a709e8a87929d93a7efdb95905d024R189


### Checklist

Delete any items that are not applicable to this PR.

- ~Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)~
- ~[Documentation](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#writing-documentation) was added for features that require explanation or tutorials~
- ~This was checked for [keyboard-only and screenreader accessibility](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility#Accessibility_testing_checklist)~
- ~This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server)~
- ~This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)~

### For maintainers
